epilogue-runtime/src/main/java/edu/wpi/first/epilogue/logging/FileBackend.java

```diff
@@ -185,4 +183,186 @@ public <P, M extends ProtoMessage<M>> void log(String identifier, P value, Proto
 
     ((ProtobufLogEntry<P>) m_entries.get(identifier)).append(value);
   }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public <P, M extends ProtoMessage<M>> void log(
+      String identifier, P value, Protobuf<P, M> proto, LogMetadata metadata) {
+    // DataLog.addSchema has checks that we're able to skip, avoiding allocations
+    if (m_seenProtos.add(proto)) {
+      m_dataLog.addSchema(proto);
+    }
+
+    if (!m_entries.containsKey(identifier)) {
+      var entry = ProtobufLogEntry.create(m_dataLog, identifier, proto);
+      if (!metadata.dependencies().isEmpty()) {
+        var dependenciesJson =
+            metadata.dependencies().stream()
+                .map(d -> '"' + d + '"')
+                .collect(Collectors.joining(", ", "[", "]"));
+        String metadataJson =
+            """
+            { "dependencies": %s }
+            """
+                .trim()
+                .formatted(dependenciesJson);
```

Comment:
This JSON generation is duplicated in the NT backend, too. Let's move it to a `LogMetadata.toJson()` method so it's reusable.